File: COA software/frontend/src/types/trainImpact.ts
export type TrainImpactLevel = 'NO_IMPACT' | 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL'

export interface AffectedTrainDetail {
  train_id: string
  train_number: string
  train_name: string
  train_type: string
  direction: 'UP' | 'DOWN'
  scheduled_entry: string
  scheduled_exit: string
  overlap_minutes: number
  estimated_delay_minutes: number
  maximum_delay_minutes: number
  priority_label: string
  impact_level: TrainImpactLevel
  passengers_affected: number
  reason: string
}

export interface AlternativeWindow {
  start_time: string
  end_time: string
  start_datetime: string
  end_datetime: string
  duration_minutes: number
  affected_trains: number
  expected_delay_minutes: number
  impact_score: number
  impact_level: TrainImpactLevel
  feasible: boolean
  reason: string
}

export interface TrainImpactSummary {
  affected_trains: number
  expected_delay_minutes: number
  maximum_delay_minutes: number
  passenger_trains: number
  goods_trains: number
  express_trains: number
  superfast_trains: number
  special_trains: number
  maintenance_trains: number
  up_trains: number
  down_trains: number
  total_passengers_estimated: number
  highest_priority: string
  impact_score: number
  operational_impact: TrainImpactLevel
  is_acceptable: boolean
}

export interface TrainImpactData {
  corridor_id: string
  corridor_name?: string
  start_time: string
  end_time: string
  start_datetime: string
  end_datetime: string
  duration_minutes: number
  block_id?: string
  block_code?: string
  summary: TrainImpactSummary
  trains: AffectedTrainDetail[]
  explanation_bullets: string[]
  recommendation: string
  alternatives: AlternativeWindow[]
  method: string
}

export interface TrainImpactRequestPayload {
  corridor_id: string
  start_time: string // ISO string
  end_time: string   // ISO string
  block_type?: string
  isolation_required?: boolean
}

File: COA software/backend/app/schemas/train_impact.py
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


class TrainImpactRequest(BaseModel):
    corridor_id: str = Field(..., description="Corridor UUID")
    start_time: datetime = Field(..., description="Proposed block start datetime")
    end_time: datetime = Field(..., description="Proposed block end datetime")
    block_type: Optional[str] = Field("MAINTENANCE", description="Block category")
    isolation_required: Optional[bool] = Field(False, description="Traction/OHE isolation flag")


class AffectedTrainDetail(BaseModel):
    train_id: str
    train_number: str
    train_name: str
    train_type: str
    direction: str
    scheduled_entry: str
    scheduled_exit: str
    overlap_minutes: int
    estimated_delay_minutes: float
    maximum_delay_minutes: float
    priority_label: str
    impact_level: str
    passengers_affected: int
    reason: str


class AlternativeWindow(BaseModel):
    start_time: str
    end_time: str
    start_datetime: str
    end_datetime: str
    duration_minutes: int
    affected_trains: int
    expected_delay_minutes: float
    impact_score: float
    impact_level: str
    feasible: bool
    reason: str


class TrainImpactSummary(BaseModel):
    affected_trains: int
    expected_delay_minutes: float
    maximum_delay_minutes: float
    passenger_trains: int
    goods_trains: int
    express_trains: int
    superfast_trains: int
    special_trains: int
    maintenance_trains: int
    up_trains: int
    down_trains: int
    total_passengers_estimated: int
    highest_priority: str
    impact_score: float
    operational_impact: str
    is_acceptable: bool


class TrainImpactData(BaseModel):
    corridor_id: str
    corridor_name: Optional[str] = None
    start_time: str
    end_time: str
    start_datetime: str
    end_datetime: str
    duration_minutes: int
    block_id: Optional[str] = None
    block_code: Optional[str] = None
    summary: TrainImpactSummary
    trains: List[AffectedTrainDetail] = []
    explanation_bullets: List[str] = []
    recommendation: str
    alternatives: List[AlternativeWindow] = []
    method: str = "Synthetic Simulation — Baseline Operational Delay Model"


class TrainImpactResponse(BaseModel):
    success: bool = True
    data: TrainImpactData
    message: str = "Train impact calculated successfully"
